### PR TITLE
Manage groups, Share studies

### LINKED ISSUE #1519: Can edit study details also inside study, not only on the dashboard view

### LINKED ISSUE #1535: files with no extension cannot be drag/dropped onto the workbench
Use-case:
as a typical linux user I have many files with no extensions. I want to drag-drop the file onto the workbench

Expected result:
drag/drop the file succeeds

Actual result:
Get ```only one file is accepted``` error

### PR DESCRIPTION
**Pull request by odeimaiz**
### Backend: #1511
- adds groups endpoint to list, create, get, modify, delete groups  through /groups endpoint
- a user in a group has read/write/delete access rights (e.g. member,manager,administrator)
- add/list/modify/delete users in groups through /groups/{gid}/users endpoint
- allow sharing of projects using access_rights defining for each group id the specific read/write/delete rights
- migrates postgres DB accordingly (migrates old access_rights syntax for projects, deprecates user_to_projects table, adds groups thumbnails, )
### Frontend:
- New page in Preferences for managing/displaying Organizations and its members
- Testers can create Organizations
- Organization Managers can invite (actually add) osparc existing members by providing their emails
- Organization Managers can remove members
- Organization Managers can promote members to Manager
- Organization SuperManagers can edit organization details
- Study owners can share the same instance of a study with Organizations and/or Organization Members (Collaborators)
- Study owners can make other collaborators Owner
- Study owners can remove collaborators

#### Backend:
- adds a decorator for checking role permissions
- fixing some issues 
- docs of how to use sc-pg migration capabilities
#### Frontend:
- Can edit study details also from inside study
- Allow dropping files with no extension
- Show hand cursor over dragable NodeUI ports
- Show move cursor over NodeUI capationbar
- Fix multiFilePicker bug
- Switch button for changing theme

## Related issue number
related ITISFoundation/osparc-issues#22
closes #1519
closes #1535 
closes ITISFoundation/osparc-issues#244

## How to test
```
cd service/web/server
make tests
```
Manage Organizations and Members
image

Share studies
image

## Checklist

- [x] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [ ] Documentation reflects the changes
- [ ] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

**Comment by GitHK** [comment]
👍 for #1535 I really wanted that

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py [hunk 1]
```diff
@@ -78,28 +80,53 @@ async def listen(app: web.Application):
             task_output = node_data["outputs"]
             node_id = node_data["node_id"]
             project_id = node_data["project_id"]
+            # FIXME: we do not know who triggered these changes. we assume the user had the rights to do so
+            # therefore we'll use the prj_owner user id. This should be fixed when the new sidecar comes in
+            # and comp_tasks/comp_pipeline get deprecated.
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py:85]: looking forward :-)

File: services/web/server/src/simcore_service_webserver/projects/projects_handlers.py [hunk 2]
```diff
@@ -13,25 +13,32 @@
...
+    "prjOwner",
+    "accessRights",
+]
 # TODO: validate these against api/specs/webserver/v0/components/schemas/project-v0.0.1.json
 
 log = logging.getLogger(__name__)
 
 
 @login_required
+@permission_required("project.create")
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:34]: nice decorators!

File: services/web/server/src/simcore_service_webserver/security_decorators.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+from typing import Coroutine
+
+from aiohttp import web
+
+from .security_api import check_permission
+
+
+def permission_required(permissions: str):
+    """Decorator that checks whether user permissions are fulfilled.
+        The function will throw an exception in case of disallowance.
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/security_decorators.py:11]: MINOR: what exception?

File: services/web/server/src/simcore_service_webserver/users_exceptions.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+"""Defines the different exceptions that may arise in the projects subpackage"""
+
+
+class UsersException(Exception):
+    """Basic exception for errors raised in projects"""
+
+    def __init__(self, msg=None):
+        if msg is None:
+            msg = "Unexpected error occured in projects subpackage"
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/users_exceptions.py:9]: MINOR: one-liner tip:
```python
msg = msg or "Unexpected ... "
```
but perhaps the ``if``-statement is more readable

File: services/web/server/src/simcore_service_webserver/users_exceptions.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+"""Defines the different exceptions that may arise in the projects subpackage"""
+
+
+class UsersException(Exception):
+    """Basic exception for errors raised in projects"""
+
+    def __init__(self, msg=None):
+        if msg is None:
+            msg = "Unexpected error occured in projects subpackage"
+        super(UsersException, self).__init__(msg)
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/users_exceptions.py:10]: Python3 is less verbose regarding ``super``. According to https://docs.python.org/3/library/functions.html#super, you can write this code as
```python
def __init__(self, msg=None):
      super().__init__(msg or "Unexpected error .." )
```

same applies to all code below

File: tests/e2e/tutorials/sleepers_project_template_sql.csv [hunk 6]
```diff
@@ -1,2 +1,2 @@
...
...acf56840e"", ""output"": ""out_2""}}, ""inputNodes"": [""template-uuid-5d8a-812c-44dacf56840e""], ""outputs"": {}, ""progress"": 0, ""thumbnail"": """", ""position"": {""x"": 550, ""y"": 200}}, ""template-uuid-5065-a079-a5a0476e3c10"": {""key"": ""simcore/services/comp/itis/sleeper"", ""version"": ""1.0.0"", ""label"": ""sleeper 3"", ""inputs"": {""in_2"": {""nodeUuid"": ""template-uuid-5f7e-92b0-5a14e84401e9"", ""output"": ""out_2""}}, ""inputNodes"": [""template-uuid-5f7e-92b0-5a14e84401e9""], ""outputs"": {}, ""progress"": 0, ""thumbnail"": """", ""position"": {""x"": 420, ""y"": 400}}, ""template-uuid-559d-aa19-dc9293e10e4c"": {""key"": ""simcore/services/comp/itis/sleeper"", ""version"": ""1.0.0"", ""label"": ""sleeper 4"", ""inputs"": {""in_1"": {""nodeUuid"": ""template-uuid-5706-b741-4073a4454f0d"", ""output"": ""out_1""}, ""in_2"": {""nodeUuid"": ""template-uuid-5065-a079-a5a0476e3c10"", ""output"": ""out_2""}}, ""inputNodes"": [""template-uuid-5706-b741-4073a4454f0d"", ""template-uuid-5065-a079-a5a0476e3c10""], ""outputs"": {}, ""progress"": 0, ""thumbnail"": """", ""position"": {""x"": 800, ""y"": 300}}}",true,"{""1"": {""read"":true, ""write"":false, ""delete"":false}}"
```
Comment (pcrespov) [inline tests/e2e/tutorials/sleepers_project_template_sql.csv:2]: this is the mega example that was failing today in master??

**Comment by pcrespov** [comment]
@odeimaiz, @sanderegg , here are some alternative names for the organization roles (inspired from gitlab and other services)
- Administrator
- Maintainer
- Coordinator
- Member (you already use this)

I think these are more informative that manager or super-manager

File: services/web/client/source/class/osparc/auth/Manager.js [hunk 7]
```diff
@@ -148,6 +148,8 @@ qx.Class.define("osparc.auth.Manager", {
     __loginUser: function(profile) {
       osparc.auth.Data.getInstance().setEmail(profile.login);
       osparc.auth.Data.getInstance().setToken(profile.login);
```
Comment (pcrespov) [inline services/web/client/source/class/osparc/auth/Manager.js:150]: please just double-check the openapi specs. I think you guys removed the entry "login" in the profile

File: tests/e2e/tutorials/sleepers_project_template_sql.csv (hunk 6 above)
Comment (odeimaiz) [reply tests/e2e/tutorials/sleepers_project_template_sql.csv:2]: Nope, this is the regular 5 sleepers template

**Comment by odeimaiz** [comment]
> @odeimaiz, @sanderegg , here are some alternative names for the organization roles (inspired from gitlab and other services)
> 
> * Administrator
> * Maintainer
> * Coordinator
> * Member (you already use this)
> 
> I think these are more informative that manager or super-manager

The super-manager was replaced by Administrator. So now we have:
- Administrator
- Manager
- Member

**Comment by pcrespov** [comment]
This is not WIP, right? remove the prefix in title

**Comment by odeimaiz** [comment]
> This is not WIP, right? remove the prefix in title

Not finished yet. Some testing and different study deletion cases to be solved.

File: services/web/client/source/class/osparc/auth/Manager.js (hunk 7 above)
Comment (odeimaiz) [reply services/web/client/source/class/osparc/auth/Manager.js:150]: still there

File: api/specs/webserver/openapi-user.yaml [hunk 8]
```diff
@@ -28,6 +28,209 @@ paths:
...
+        - user
+      summary: Deletes one group
+      operationId: delete_group
+      responses:
+        "204":
+          description: group has been successfully deleted
+        default:
+          $ref: "#/components/responses/DefaultErrorResponse"
+
+  /me/groups/{gid}/users:
```
[the remark this answers is not in the document]
Comment (ignapas) [reply api/specs/webserver/openapi-user.yaml]: :S sorry... it was not very critical either... but cool it's done :)

File: services/web/client/source/class/osparc/dashboard/OrganizationEditor.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+   License:
+     MIT: https://opensource.org/licenses/MIT
+
+   Authors:
+     * Odei Maiz (odeimaiz)
+
+************************************************************************ */
+
+qx.Class.define("osparc.dashboard.OrganizationEditor", {
```
Comment (ignapas) [inline services/web/client/source/class/osparc/dashboard/OrganizationEditor.js:18]: what about ``JsonSchemaForm``...? let's reuse it!

Comment (ignapas) [reply services/web/client/source/class/osparc/dashboard/OrganizationEditor.js:18]: it's the perfect opportunity to make it really useful. this guy has validation for free... and no need to write more code for each new form we want... just define it with jsonschema (easy to change too)

Comment (odeimaiz) [reply services/web/client/source/class/osparc/dashboard/OrganizationEditor.js:18]: Will take it into account for the next form 👍

**Comment by odeimaiz** [comment]
> My suggestions:
> * No need for a full section (with redundant title) for the Create Organisation button
> * Remove general padding for sections (specially for the organizations list)
> * 89 members?
> * In the edit organization dialog, the cancel button doesn't do anything
> * I can create organizations with an empty title... This should be controlled both server and client side.
> * Remove ultra-random default thumbnail?
>   The invite function is not very clear... I can remove myself (the creator) from the organization and then I cannot add me again. I have more stuff, maybe later we can spend some quality time

All done except the padding one

File: services/web/client/source/class/osparc/component/filter/Organizations.js [hunk 10]
```diff
@@ -27,7 +27,7 @@ qx.Class.define("osparc.component.filter.Organizations", {
    * @extends osparc.component.filter.TagsFilter
    */
   construct: function() {
-    this.base(arguments, this.tr("My Organizations"), "organizations", "organizations");
+    this.base(arguments, this.tr("Select Organization"), "organizations", "organizations");
```
Comment (GitHK) [inline services/web/client/source/class/osparc/component/filter/Organizations.js:30]: 👍

File: services/web/client/source/class/osparc/auth/Manager.js [hunk 11]
```diff
@@ -148,6 +148,8 @@ qx.Class.define("osparc.auth.Manager", {
...
+      osparc.auth.Data.getInstance().setUserId(profile.id);
+      osparc.auth.Data.getInstance().setGroupId(profile["groups"]["me"]["gid"]);
```
Comment (pcrespov) [inline services/web/client/source/class/osparc/auth/Manager.js:152]: MINOR: why are you not using ``profile.groups.me.gid`` here as you do with the others?

**Pull request merged**